rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,359 @@
+used to control which MCP servers a role can access based on the labels of the
+MCP server apps.
+
+A new role option `mcp_tools` is introduced to apply fine-grained control on
+which tools from the MCP server should be allowed. Similar to
+`kubernetes_resources.name`, entries in `mcp_tools` also support glob and regex
+representations to make whitelisting or blacklisting easier. If `mcp_tools` is not
+specified for a role, all tools are allowed by default. This new option
+`mcp_tools` will be enforced per JSON-RPC `tool/call` call, which will be
+detailed in the next section.
+
+The procedure to access an MCP server from `tsh` is the same as any other app --
+a user has to obtain a user certificate with `route_to_app` for the TLS routing
+handshake, and the backend validates the identity then routes to the appropriate
+app service based on this certificate.
+
+#### Implementation - MCP server handler with JSON-RPC
+
+MCP server handler will be placed in App service's connection handler. The MCP
+server handler will receive a raw connection once the connection is authorized
+and confirmed to be an MCP app. It is expected [the MCP protocol in JSON-RPC
+format](https://modelcontextprotocol.io/specification/2025-03-26) will be
+transferred within this connection.
+
+Since using `stdio`, a `exec.Command` defined by the app definition is executed
+to start an MCP server, for each authorized incoming client connection. Then
+stdin and stdout are proxied between the client connection and the launched MCP
+server. Stderr from the launched MCP server will be logged at TRACE level in
+Teleport logs.
+
+Incoming bytes from the client do get the following treatments before they are
+forwarded to the launched MCP server:
+- Parsed into MCP/JSON-RPC messages. 
+- If the message is a `tools/call` request, use the access-checker to confirm
+  whether the identity has access to the requested tool.
+  - If the `tools/call` is access denied, make a text result detailing this
+    error and send it back to the client connection. Do not forward the request to
+    the launched MCP server in this case.
+- Record audit events for eligible messages.
+
+The launched MCP server should be stopped once the connection ends.
+
+#### Implementation - "tsh mcp" commands
+
+A suite of `tsh mcp` commands are added on the client side:
+- `tsh mcp ls`: lists MCP server apps, with slightly different columns from `tsh
+  app ls`.
+- `tsh mcp connect`: command to be called by the AI tool to launch an MCP server
+  through Teleport. User cert with `route_to_app` is automatically obtained if
+  the app hasn't been logged in yet. stdin and stdout of this process are
+  proxied to Teleport backend. Any local tsh logs are sent to stderr.
+- `tsh mcp config`: shows a sample JSON configuration that launches the MCP
+  server with `tsh mcp connect` for Claude Desktop. (Most AI tools use the same
+  JSON format as well.)
```

Comment:
Perhaps `tsh mcp login` would be useful too?
- requests app cert with specific parameters
- calls `tsh mcp config` automatically

Reply:
dropped `tsh mcp config`, and now just use `tsh mcp login` instead.